livechatsnew/consumers.py

```diff
@@ -0,0 +1,115 @@
+import json
+
+import pydantic
+from bson import ObjectId
+from datetime import datetime
+from django.core.cache import cache
+from collections import defaultdict
+from channels.db import database_sync_to_async
+from channels.generic.websocket import AsyncWebsocketConsumer
+from livechatsnew.utils import mongo_conversations
+from livechatsnew.schemas import Message
+from pydantic import parse_obj_as
+
+
+class ChatConsumer(AsyncWebsocketConsumer):
+    room_connection_counts = defaultdict(lambda: 0)
+
+    @database_sync_to_async
+    def get_conversation(self):
+        conversations = mongo_conversations()
+        conv = conversations.find_one({"_id": ObjectId(self.room_name)})
+        initiator = conv.get("initiator_id")
+        receiver = conv.get("receiver_id")
+        return conv, initiator, receiver
+
+    @database_sync_to_async
+    def update_conversation(self, messages):
+        """Function for updating live-chat messages data"""
+        conversations = mongo_conversations()
+        conversations.update_one({"_id": ObjectId(self.room_name)}, {"$push": {"messages": messages}})
+
+    async def create_message_data(self, message):
+        """Function for creating messages available for caching and sending"""
+        conversation = await self.get_conversation()
+        sender = self.scope["user"]
+        message = {
+            "msg_text": message["message"],
+            "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
+            "sender_id": sender.id,
+            "sender_data": str(sender),
+            "conversation_id": str(conversation[0].get("_id")),
+
+        }
+        return message
+
+    async def cache_data(self, message):
+        """Function for adding messages and joining data to cache for 10 hours"""
+        _message = await self.create_message_data(message)
+        chat_value = cache.get(_message["conversation_id"])
+        if chat_value:
+            chat_value = str(_message) + "," + chat_value
+        else:
+            chat_value = str(_message)
+        cache.set(_message["conversation_id"], chat_value, 36000)
+
+    async def connect(self):
+        """
+        Connecting to the WS, also adding number to "room_connection_counts"
+         for counting the number of active users in a channel
+         """
+        self.room_name = self.scope["url_route"]["kwargs"]["room_name"]
+        self.room_group_name = f"chat_{self.room_name}"
+        conversation, initiator, receiver = await self.get_conversation()
+        participants_of_conv = [initiator, receiver]
+
+        current_user = self.scope['user']
+
+        if current_user.is_authenticated and current_user.id in participants_of_conv:
+            await self.channel_layer.group_add(
+                self.room_group_name, self.channel_name
+            )
+
+            await self.accept()
+            self.room_connection_counts[self.room_name] += 1
```

Comment:
Consider edge cases where the same user might open multiple connections to the same room, or error scenarios where connections might not be cleanly closed, leading to inaccurate counts